tests/modules/saml/src/Controller/ServiceProviderTest.php

```diff
@@ -185,6 +208,247 @@ public function testWithAuthIdWithoutEntity(): void
+        }
+    }
+
+    /**
+     * @return array
+     */
+    public static function loginAuthenticatedDataProvider(): array
+    {
+        return [
+            'no query params' => [
+                [], // If no query params are provided, it should default to the RelayState
+                [
+                    'ReturnTo' =>  self::RELAY_STATE,
+                ],
+                false
+            ],
+            'entityID=example.edu' => [
+                [
+                    'entityID' => 'example.edu',
+                ],
+                [
+                    'ReturnTo' =>  self::RELAY_STATE,
+                ],
+                false
+            ],
+            'entityID=example.edu&forceAuthn=false&isPassive=false' => [
+                [
+                    'entityID' => 'example.edu',
+                    'forceAuthn' => false,
+                    'isPassive' => false
+                ],
+                [
+                    'ReturnTo' =>  self::RELAY_STATE,
+                ],
+                false
+            ],
+            'entityID=other.edu' => [
+                [
+                    'entityID' => 'other.edu',
+                ],
+                [
+                    'ReturnTo' =>  self::RELAY_STATE,
+                    'saml:idp' => 'other.edu'
+                ],
+                true
+            ],
+            'forceAuthn=true&target=/some/url' => [
+                [
+                    'target' => '/some/url',
+                    'forceAuthn' => true
+                ],
+                [
+                    'ReturnTo' => 'http://localhost/some/url',
+                    'ForceAuthn' => true
+                ],
+                true
+            ],
+            'target=https://evil.com' => [
+                [
+                    'target' => 'https://evil.com'
+                ],
+                [],
+                false,
+                true
+            ]
+        ];
+    }
+
+    /**
+     * Test Login with Session/Authenticated
+     *
+     * @dataProvider loginAuthenticatedDataProvider
+     *
+     * @param   array  $queryParameters
+     * @param   array  $options
+     * @param   bool   $expectLoginCalled
+     * @param   bool   $expectingException
+     *
+     * @return void
+     */
+    public function testLoginHandleAuthenticated(
+        array $queryParameters,
+        array $options,
+        bool $expectLoginCalled,
+        bool $expectingException = false
+    ): void {
+        $request = Request::create(
+            '/sp/login/phpunit',
+            'GET',
+            $queryParameters
+        );
+
+        $info = ['AuthId' => 'phpunit'];
+        $config =
+            [
+                'saml:SP',
+                'entityID' => 'urn:x-simplesamlphp:example-sp',
+                'RequestInitiation' => true,
+                'RelayState' => self::RELAY_STATE,
+            ];
+
+        $spSource = new class ($info, $config) extends Source\SP {
```

Comment:
I'm curious about the need for using an anonymous subclass of `Source\SP` rather than using it directly